### DISCUSSION TITLE
Is there any way to prepopulate the querybuilder with the fields and values if query is passed as input.

### DISCUSSION THREAD

**Discussion by jaganryali** [react-querybuilder/react-querybuilder#746] [Q&A] [answered]
Hi Everyone,

I'm trying to Save query generated by selecting rules including fields and values in DB.
When user wants to update that query again I don't want user to start it from adding all rules again, Looking if there is a way which it accepts generated query and then pre-populates the rules, fields and value selection.

**Comment by jakeboone02** [comment, marked as answer]
The `defaultQuery` prop will set the initial query. To use the component as a controlled component, use the `query` prop instead.

**Reply by jaganryali** [reply]
Hi @jakeboone02 ,

I have a scenario like, user will choose the rules and then I'm persisting the formatQuery output. When I what to use prepopulate the component to use as controlled component i'm getting error like



Cannot create property 'id' on string '{"combinator":"and","rules":[{"field":"firstname","operator":"!=","valueSource":"field","value":"lastname"}]}'



const initialQuery:RuleGroupType = {
    combinator: 'and',
    rules: [],
  };

  const [query, setQuery] = useState<RuleGroupType>(queryProps ?? initialQuery);

queryProps is persisted value.

**Reply by jaganryali** [reply]
Got the issue, I'm persisting the result of component by storing the result of formatQuery('json_without_ids').
Result is converted to string and storing, When it is pulled from DB, even though when we parse the string to Object it is still be like string rather than converting to RuleGroupType. 
I had to convert to unknown and parse it again and cast it to RuleGroupType to feed to QueryBuilder
